### PR TITLE
feat: decouple DID from all other pallets to support independent updates

### PR DESCRIPTION
**Pull request by Diiaablo95**
## fixes KILTProtocol/ticket#1369

This PR refactors specifically the delegation pallet, which before was relying on the DID pallet to perform delegate's signature verification. Now, the verification operation can be provided in the pallet's configuration by implementing a new trait. This is done in the runtimes.
All mentions of DIDs have been removed from ctype, attestation, and delegation pallets also for test cases, as now a simple `EnsureSigned` origin is used to test those.

The longer-term goal that includes this PR is to allow a mashnet node release that does not completely break the SDK, as all KILT operations but DID management do not require now any DID to be present on chain (as can be tested via Polkadot-JS apps by creating a CTYPE via the ctype pallet and not the DID pallet anymore).

> Some of the changes in this PR come from [Albi's brench](https://github.com/KILTprotocol/mashnet-node/tree/aw-update-dependencies), which I had to integrate into mine to compile successfully 🎉.

## How to test with Cargo

Run `cargo test --all --all-features`.

## How to test with Polkadot-JS apps

Start the node, and try to create a CTYPE, an attestation and/or a delegation via the respective pallets. They will work! 😃

### Custom types

The following types are now required for the new chain version. As there is no practical difference between `AccountId` and `DidIdentifier` at the moment, the old ones still one work, but I think it is better to explicitly refer to them for what they really are in practice.
Nevertheless, I am not sure if we should open another PR for these types, and if so what version would they be. So I might need some guidance there.

<details>
  <summary>JS-Types</summary>
  
  ```json
{
  "AccountInfo": "AccountInfoWithTripleRefCount",
  "Address": "MultiAddress",
  "AmountOf": "i128",
  "Balance": "u128",
  "BlockNumber": "u64",
  "Index": "u64",
  "LookupSource": "MultiAddress",
  "CtypeCreatorOf": "AccountId",
  "CtypeHashOf": "Hash",
  "ClaimHashOf": "Hash",
  "AttesterOf": "AccountId",
  "AttestationDetails": {
    "ctypeHash": "CtypeHashOf",
    "attester": "AttesterOf",
    "delegationId": "Option<DelegationNodeIdOf>",
    "revoked": "bool"
  },
  "Permissions": "u32",
  "DelegationNodeIdOf": "Hash",
  "DelegatorIdOf": "AccountId",
  "DelegateSignatureTypeOf": "Vec<u8>",
  "DelegationRoot": {
    "ctypeHash": "CtypeHashOf",
    "owner": "DelegatorIdOf",
    "revoked": "bool"
  },
  "DelegationNode": {
    "rootId": "DelegationNodeIdOf",
    "parent": "Option<DelegationNodeIdOf>",
    "owner": "DelegatorIdOf",
    "permissions": "Permissions",
    "revoked": "bool"
  },
  "KeyIdOf": "Hash",
  "DidIdentifierOf": "AccountId",
  "AccountIdentifierOf": "AccountId",
  "BlockNumberOf": "BlockNumber",
  "DidCallableOf": "Call",
  "DidVerificationKey": {
    "_enum": {
      "Ed25519": "[u8; 32]",
      "Sr25519": "[u8; 32]"
    }
  },
  "DidEncryptionKey": {
    "_enum": {
      "X25519": "[u8; 32]"
    }
  },
  "DidPublicKey": {
    "_enum": {
      "PublicVerificationKey": "DidVerificationKey",
      "PublicEncryptionKey": "DidEncryptionKey"
    }
  },
  "DidVerificationKeyRelationship": {
    "_enum": [
      "Authentication",
      "CapabilityDelegation",
      "CapabilityInvocation",
      "AssertionMethod"
    ]
  },
  "DidSignature": {
    "_enum": {
      "Ed25519": "Ed25519Signature",
      "Sr25519": "Sr25519Signature"
    }
  },
  "DidError": {
    "_enum": {
      "StorageError": "StorageError",
      "SignatureError": "SignatureError",
      "UrlError": "UrlError",
      "InternalError": "Null"
    }
  },
  "StorageError": {
    "_enum": {
      "DidAlreadyPresent": "Null",
      "DidNotPresent": "Null",
      "DidKeyNotPresent": "DidVerificationKeyRelationship",
      "VerificationKeyNotPresent": "Null",
      "CurrentlyActiveKey": "Null",
      "MaxTxCounterValue": "Null"
    }
  },
  "SignatureError": {
    "_enum": [
      "InvalidSignatureFormat",
      "InvalidSignature",
      "InvalidNonce"
    ]
  },
  "KeyError": {
    "_enum": [
      "InvalidVerificationKeyFormat",
      "InvalidEncryptionKeyFormat"
    ]
  },
  "UrlError": {
    "_enum": [
      "InvalidUrlEncoding",
      "InvalidUrlScheme"
    ]
  },
  "DidPublicKeyDetails": {
    "key": "DidPublicKey",
    "blockNumber": "BlockNumberOf"
  },
  "DidDetails": {
    "authenticationKey": "KeyIdOf",
    "keyAgreementKeys": "BTreeSet<KeyIdOf>",
    "delegationKey": "Option<KeyIdOf>",
    "attestationKey": "Option<KeyIdOf>",
    "publicKeys": "BTreeMap<KeyIdOf, DidPublicKeyDetails>",
    "endpointUrl": "Option<Url>",
    "lastTxCounter": "u64"
  },
  "DidCreationOperation": {
    "did": "DidIdentifierOf",
    "newAuthenticationKey": "DidVerificationKey",
    "newKeyAgreementKeys": "BTreeSet<DidEncryptionKey>",
    "newAttestationKey": "Option<DidVerificationKey>",
    "newDelegationKey": "Option<DidVerificationKey>",
    "newEndpointUrl": "Option<Url>"
  },
  "DidUpdateOperation": {
    "did": "DidIdentifierOf",
    "newAuthenticationKey": "Option<DidVerificationKey>",
    "newKeyAgreementKeys": "BTreeSet<DidEncryptionKey>",
    "attestationKeyUpdate": "DidVerificationKeyUpdateAction",
    "delegationKeyUpdate": "DidVerificationKeyUpdateAction",
    "publicKeysToRemove": "BTreeSet<KeyIdOf>",
    "newEndpointUrl": "Option<Url>",
    "txCounter": "u64"
  },
  "DidVerificationKeyUpdateAction": {
    "_enum": {
      "Ignore": "Null",
      "Change": "DidVerificationKey",
      "Delete": "Null"
    }
  },
  "DidDeletionOperation": {
    "did": "DidIdentifierOf",
    "txCounter": "u64"
  },
  "DidAuthorizedCallOperation": {
    "did": "DidIdentifierOf",
    "txCounter": "u64",
    "call": "DidCallableOf"
  },
  "HttpUrl": {
    "payload": "Text"
  },
  "FtpUrl": {
    "payload": "Text"
  },
  "IpfsUrl": {
    "payload": "Text"
  },
  "Url": {
    "_enum": {
      "Http": "HttpUrl",
      "Ftp": "FtpUrl",
      "Ipfs": "IpfsUrl"
    }
  },
  "LockedBalance": {
    "block": "BlockNumber",
    "amount": "Balance"
  },
  "BalanceOf": "Balance",
  "RoundInfo": {
    "current": "SessionIndex",
    "first": "BlockNumber",
    "length": "BlockNumber"
  },
  "OrderedSet": "Vec<Stake>",
  "Stake": {
    "owner": "AccountId",
    "amount": "Balance"
  },
  "TotalStake": {
    "collators": "Balance",
    "delegators": "Balance"
  },
  "InflationInfo": {
    "collator": "StakingInfo",
    "delegator": "StakingInfo"
  },
  "StakingInfo": {
    "maxRate": "Perquintill",
    "rewardRate": "RewardRate"
  },
  "RewardRate": {
    "annual": "Perquintill",
    "perBlock": "Perquintill"
  },
  "Delegator": {
    "delegations": "Vec<Stake>",
    "total": "Balance"
  },
  "CollatorSnapshot": {
    "stake": "Balance",
    "delegators": "Vec<Stake>",
    "total": "Balance"
  },
  "Collator": {
    "id": "AccountId",
    "stake": "Balance",
    "delegators": "Vec<Stake>",
    "total": "Balance",
    "state": "CollatorStatus"
  },
  "CollatorStatus": {
    "_enum": {
      "Active": "Null",
      "Leaving": "SessionIndex"
    }
  }
}
  ```
</details>

## Checklist:

- [x] I have verified that the code works
  - [x] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [x] I have verified that the code is easy to understand
  - [x] If not, I have left a well-balanced amount of inline comments
- [ ] This PR does not introduce new custom types
  - [ ] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

File: pallets/delegation/src/lib.rs [hunk 1]
```diff
@@ -211,32 +219,19 @@ pub mod pallet {
 			parent_id: Option<DelegationNodeIdOf<T>>,
 			delegate: DelegatorIdOf<T>,
 			permissions: Permissions,
-			delegate_signature: did::DidSignature,
+			delegate_signature: Vec<u8>,
```
Comment (weichweich) [inline pallets/delegation/src/lib.rs:222]: maybe use the associate type  `DelegationSignatureVerification::Signature` here?

**Review by weichweich — APPROVED** [review]
It's a bit messy because it includes my PR. 😓 But LGTM! nice work. 👍🏻

**Review by wischli — CHANGES_REQUESTED** [review]
Also unhappy about the mess of @weichweich's changes and this one. I would propose to merge this PR into `aw-update-dependencies`.

If this should be merged urgently, I would propose to remove the chain specs from this PR, at least.

**Comment by Diiaablo95** [comment]
Closed in favor of https://github.com/KILTprotocol/mashnet-node/pull/179.

File: pallets/delegation/src/lib.rs (hunk 1 above)
Comment (Diiaablo95) [reply pallets/delegation/src/lib.rs]: Fixed in https://github.com/KILTprotocol/mashnet-node/pull/179/commits/6d1d7a766b55e1f2ab874caf87090b7db2e6e300.

**Pull request closed without merging**